### ISSUE TITLE
Incorrect handle position after node rotation

### ISSUE THREAD

**Issue by zaxarhic2134** [wbkd/react-flow#3479]
### Describe the Bug

During development, I encountered a problem regarding the position of the handle after rotating the element; after rotating the element by 90 degrees when connecting a line, it is drawn incorrectly, and this is due to the fact that when the entire node is rotated, the handle also changes its position, but its position in NodeInternals is not changes, thereby adding an incomprehensible offset

### Your Example Website or App

https://codesandbox.io/s/competent-dubinsky-yd4kqr?file=/src/FourWayPipe.js

### Steps to Reproduce the Bug or Issue

1. Connect nodes
2. Make sure the display is correct
3. Delete a connecting line by clicking on it and pressing BackSpace
4. Rotate one of the nodes 90 degrees connect them

### Expected behavior

I expected that when turning everything would connect correctly, but unfortunately the connections do not occur correctly, I also checked official documentation and found the same error there

### Screenshots or Videos

![image](https://github.com/wbkd/react-flow/assets/91330925/d5858662-0693-4363-87f5-4ab86a4138e7)
![image](https://github.com/wbkd/react-flow/assets/91330925/d75cc5cf-9b81-4f17-8094-33aaa5779955)

### Platform

- OS: Linux
- Browser: Chrome
- Version: 11.8.3

### Additional context

I tried to dynamically change the position of the handle when rotating the element in two ways, but this was also unsuccessful
`

    useEffect(() => {
                const pos = [Position.Top, Position.Left, Position.Bottom, Position.Right]
                const index = rotation/90
                const angle = Math.abs(index) === 4 ? 0 : Math.abs(index)
                const arr = pos.splice(-angle).concat(pos)
                setPosition(arr)
                updateNodeInternals(id)
            }, [rotation])

    const HandlePosition = observer(({rotation}) => {
        const pos = [Position.Top, Position.Left, Position.Bottom, Position.Right]
        const index = rotation/90
        const angle = Math.abs(index) === 4 ? 0 : Math.abs(index)
        const position = pos.splice(-angle).concat(pos)
        console.log(position[0])
        updateNodeInternals(id)

        return (
            <>
                <Handle
                    id="fwp1"
                    type="source"
                    position={position[0]}
                    style={{background: '#555' }}
                />
                <Handle
                    id="fwp2"
                    type="source"
                    position={position[1]}
                    style={{background: '#555' }}
                />
                <Handle
                    id="fwp3"
                    type="source"
                    position={position[2]}
                    style={{background: '#555' }}
                />
                <Handle
                    id="fwp4"
                    type="source"
                    position={position[3]}
                    style={{background: '#555' }}
                />
            </>
        )
    })
`

When implementing these methods, I checked the correctness of the work by displaying sourcePosition and targetPosition in the console, which are needed to determine the path of the drawn line, but the display there is also not correct

**Comment by LeandroParisi** [comment]
On your code sandbox it seems like it is not possible to rotate the elements, even though the rotate handler is in place when I click and try dragging it to rotate it does not work.

**Comment by LeandroParisi** [comment]
@moklick I was thinking about trying to fix this issue in order to contribute to the project.
I contacted you last friday on discord about contributing.
Do you feel I could tackle this one, and, if so: Do you have any insight which might help me out here about this issue?

**Comment by LeandroParisi** [comment]
> ### Describe the Bug
> During development, I encountered a problem regarding the position of the handle after rotating the element; after rotating the element by 90 degrees when connecting a line, it is drawn incorrectly, and this is due to the fact that when the entire node is rotated, the handle also changes its position, but its position in NodeInternals is not changes, thereby adding an incomprehensible offset
> 
> ### Your Example Website or App
> https://codesandbox.io/s/competent-dubinsky-yd4kqr?file=/src/FourWayPipe.js
> 
> ### Steps to Reproduce the Bug or Issue
> 1. Connect nodes
> 2. Make sure the display is correct
> 3. Delete a connecting line by clicking on it and pressing BackSpace
> 4. Rotate one of the nodes 90 degrees connect them
> 
> ### Expected behavior
> I expected that when turning everything would connect correctly, but unfortunately the connections do not occur correctly, I also checked official documentation and found the same error there
> 
> ### Screenshots or Videos
> ![image](https://user-images.githubusercontent.com/91330925/272841313-d5858662-0693-4363-87f5-4ab86a4138e7.png) ![image](https://user-images.githubusercontent.com/91330925/272841520-d75cc5cf-9b81-4f17-8094-33aaa5779955.png)
> 
> ### Platform
> * OS: Linux
> * Browser: Chrome
> * Version: 11.8.3
> 
> ### Additional context
> I tried to dynamically change the position of the handle when rotating the element in two ways, but this was also unsuccessful `
> 
> ```
> useEffect(() => {
>             const pos = [Position.Top, Position.Left, Position.Bottom, Position.Right]
>             const index = rotation/90
>             const angle = Math.abs(index) === 4 ? 0 : Math.abs(index)
>             const arr = pos.splice(-angle).concat(pos)
>             setPosition(arr)
>             updateNodeInternals(id)
>         }, [rotation])
> 
> const HandlePosition = observer(({rotation}) => {
>     const pos = [Position.Top, Position.Left, Position.Bottom, Position.Right]
>     const index = rotation/90
>     const angle = Math.abs(index) === 4 ? 0 : Math.abs(index)
>     const position = pos.splice(-angle).concat(pos)
>     console.log(position[0])
>     updateNodeInternals(id)
> 
>     return (
>         <>
>             <Handle
>                 id="fwp1"
>                 type="source"
>                 position={position[0]}
>                 style={{background: '#555' }}
>             />
>             <Handle
>                 id="fwp2"
>                 type="source"
>                 position={position[1]}
>                 style={{background: '#555' }}
>             />
>             <Handle
>                 id="fwp3"
>                 type="source"
>                 position={position[2]}
>                 style={{background: '#555' }}
>             />
>             <Handle
>                 id="fwp4"
>                 type="source"
>                 position={position[3]}
>                 style={{background: '#555' }}
>             />
>         </>
>     )
> })
> ```
> 
> `
> 
> When implementing these methods, I checked the correctness of the work by displaying sourcePosition and targetPosition in the console, which are needed to determine the path of the drawn line, but the display there is also not correct

On your code sandbox it seems like it is not possible to rotate the elements, even though the rotate handler is in place when I click and try dragging it to rotate it does not work.

**Comment by zaxarhic2134** [comment]
> On your code sandbox it seems like it is not possible to rotate the elements, even though the rotate handler is in place when I click and try dragging it to rotate it does not work.

The angle of rotation of the element is strictly 90 degrees. This is done to avoid problems with handlers

**Comment by moklick** [comment]
what is not working here? When I rotate a node, the connections still look fine to me and I can still create new connections.

**Comment by zaxarhic2134** [comment]
> what is not working here? When I rotate a node, the connections still look fine to me and I can still create new connections.


In my example, the connection line is displayed incorrectly after rotation as shown in the screenshots, and this can also be seen in the sandbox. And if we are talking about official documentation, then this can be seen in the screenshot below:

![image](https://github.com/wbkd/react-flow/assets/91330925/cced0477-30c7-4c18-b522-4b4ace177fcd)
![image](https://github.com/wbkd/react-flow/assets/91330925/c42d4291-5f7e-448a-9daf-af59ff86ccdf)

Or here from the new version, but for a clearer explanation I changed the edge type in the sandbox

![image](https://github.com/wbkd/react-flow/assets/91330925/2d7011d4-9b40-43a5-8b59-3fe62466ce75)
![image](https://github.com/wbkd/react-flow/assets/91330925/105ad893-498f-4a38-9b74-02fd7279057c)


In this case, we see that in the normal position the line comes out from below, and the vector of its beginning is directed downward, but when turning, the line already comes out from the left, but the vector of its beginning is still directed downward, although it should be directed to the left, and apparently because of this in my In the example, the connection line is not displayed correctly

**Comment by moklick** [comment]
For this you would need to change the handle position dynamically or implement a custom edge type that could figure out the handle position on its own. Does this make sense to you?

**Comment by zaxarhic2134** [comment]
> Для этого вам нужно будет динамически изменять положение ручки или реализовать собственный тип ребра, который мог бы самостоятельно определять положение ручки. Имеет ли это смысл для вас?

Yes, this is exactly what I need, but I don’t understand how to implement it

**Comment by moklick** [comment]
You need to do some calculations that help you to find out if a handle is left, top, right or bottom. Roughly like:

1. rotate node
2. get handle positions with `getBoundingClientRect`
3. compare positions (smallest y = top, biggest y = bottom, smallest x = left, biggest x = right)  
4. assign new handle positions
